lib/cloner.ml

```diff
@@ -0,0 +1,138 @@
+      Ok { cache_dir = Some cache_dir }
+
+module Cached = struct
+  let valid_remote_name = Astring.String.map (function '/' | '@' | ':' -> '-' | x -> x)
+
+  let commit_branch_name ~commit = "commit-" ^ commit
+
+  let cache_branch_name ~ref = "branch-" ^ ref
+
+  (** Check if the cache repository is initialized and if not, initialize it *)
+  let check_duniverse_cache_directory ~repo ~remote =
+    let open Result.O in
+    let ok_file = Fpath.(repo / "duniverse_OK") in
+    OS.Path.exists ok_file >>= function
+    | true -> Ok ()
+    | false ->
+        OS.Dir.delete ~recurse:true repo >>= fun () ->
+        OS.Dir.create repo >>= fun _ ->
+        Exec.git_init_bare ~repo >>= fun () ->
+        Exec.git_remote_add ~repo ~remote_url:remote ~remote_name:"origin" >>= fun () ->
+        OS.File.write ok_file ""
+
+  (** Get the path for the given remote in the cache *)
+  let get_cache_directory ~remote cache_dir =
+    let open Result.O in
+    let remote_dir = valid_remote_name remote in
+    let repo = Fpath.(cache_dir / remote_dir) in
+    check_duniverse_cache_directory ~repo ~remote >>| fun () -> repo
+
+  let git_branch_exists_or_create ~repo ~ref ~branch =
+    let open Result.O in
+    if Exec.git_branch_exists ~repo ~branch then Ok true
+    else Exec.git_fetch_to ~repo ~remote_name:"origin" ~ref ~branch () >>| fun () -> false
+
+  let git_commit_branch_exists_or_create ~repo ~ref ~branch ~commit ~commit_branch () =
+    let open Result.O in
+    if Exec.git_branch_exists ~repo ~branch:commit_branch then Ok true
+    else
+      git_branch_exists_or_create ~repo ~ref ~branch >>= fun cached ->
+      Exec.git_branch ~repo ~ref:commit ~branch_name:commit_branch >>| fun () -> cached
+
+  (** Check if a remote with a given tag exists in the cache as a branch, and clone to cache if it
+      doesn't exist *)
+  let check_package_cache_branch ~repo ~ref ~commit () =
+    let open Result.O in
+    let commit_branch = commit_branch_name ~commit in
+    let branch = cache_branch_name ~ref in
+    git_commit_branch_exists_or_create ~repo ~ref ~branch ~commit ~commit_branch ()
+    >>| fun cached -> (commit_branch, cached)
+
+  (** Clone to output_dir using the cache *)
+  let clone_from_cache ~output_dir ~repo (cache_branch_name, cached) =
+    let open Result.O in
+    OS.Dir.delete ~recurse:true output_dir >>= fun () ->
+    Exec.git_clone ~branch:cache_branch_name ~remote:(Fpath.to_string repo) ~output_dir
+    >>| fun () -> cached
+
+  let clone_to ~output_dir ~remote ~ref ~commit cache_dir =
+    let open Result.O in
+    get_cache_directory ~remote cache_dir >>= fun repo ->
+    check_package_cache_branch ~repo ~ref ~commit () >>= clone_from_cache ~output_dir ~repo
+end
+
+module Uncached = struct
+  let warn_about_head_commit ~ref ~commit () =
+    Logs.info (fun l ->
+        l "%a is not the HEAD commit for %a anymore" Styled_pp.commit commit Styled_pp.branch ref
```

Comment:
Not a blocker for this PR, but having a warning about HEAD could be annoying if you are deliberately pinning a package. There's no way to disable this warning as far as I can tell.

Reply:
This actually is an info log, which means it is disabled by default and you'll only see it if you run with the `-v` option. I believe it can be valuable information in a verbose output as it can help you figure out why you're not in sync with the tracked branch after a pull!

What do you think?

We actually changed it in #53 I believe, it used to be a warning before!